File: test/qtsocket/comunicator.cpp
#include <QObject>
#include <QThreadPool>
#include <QByteArray>
#include <QTcpSocket>
#include <QList>
#include <QMutex>

#include "comunicator.h"

Comunicator::Comunicator(qintptr socketDescriptor, QObject *parent = nullptr) 
    : QObject(parent), socket(new QTcpSocket(this)) {
    if (!socket->setSocketDescriptor(socketDescriptor)) {
        qDebug() << "Failed to set socket descriptor.";
        return;
    }
    qDebug() << "Comunicator - initialized";
    
    connect(socket, &QTcpSocket::readyRead, this, &Comunicator::readData);
    connect(socket, &QTcpSocket::bytesWritten, this, &Comunicator::onWriteFinish);
    connect(socket, &QTcpSocket::disconnected, this, &Comunicator::onDisconnected);
}

bool Comunicator::begin() {
    socket->waitForReadyRead(3000);  // wait for data to be ready
    return socket->bytesAvailable() > 0;
}

void Comunicator::readData() {
    QByteArray data = socket->readAll();
    qDebug() << "Received data:" << data;
    emit dataReceived(data);
}

void Comunicator::sendData(const QByteArray &data) {
    socket->write(data);
}

void Comunicator::onWriteFinish(qint64 bytes) {
    qDebug() << "Data written:" << bytes;
}

void Comunicator::onDisconnected() {
    qDebug() << "Socket disconnected";
    emit requestDelete();
}

File: test/qtsocket/comunicator.py
from PyQt5.QtCore import QObject, pyqtSignal, pyqtSlot, QByteArray
from PyQt5.QtNetwork import QTcpSocket

class Comunicator(QObject):
    dataReceived = pyqtSignal(QByteArray)
    writeFinish = pyqtSignal('qint64')
    requestDelete = pyqtSignal()

    def __init__(self, sd, parent=None):
        super().__init__(parent)
        self.sd = sd
        self.socket = QTcpSocket(self)
        self.socket.readyRead.connect(self.readData)
        self.socket.bytesWritten['qint64'].connect(self.writeFinish.emit)
        self.socket.disconnected.connect(self.requestDelete.emit)
        print("Comunicator - initialized")

    @pyqtSlot()
    def begin(self):
        if not self.socket.setSocketDescriptor(self.sd):
            return False
        print('Socket descriptor sets')
        self.socket.waitForReadyRead(3000)
        return True
    
    @pyqtSlot()
    def readData(self):
        print('read Data called')
        data = self.socket.readAll()
        print("data :", data)
        self.dataReceived.emit(QByteArray(data))
        # self.requestDelete.emit()
   
    @pyqtSlot(str,QByteArray)
    def sendData(self, ch, data):
        self.socket.write(data)